crates/net/network/src/transactions.rs

```diff
@@ -524,47 +545,204 @@ where
+                let is_valid_for_version_eth68 = peer.version == EthVersion::Eth68;
+                self.transaction_fetcher.fill_request_for_peer(
+                    &mut hashes,
+                    peer_id,
+                    is_valid_for_version_eth68,
+                );
+
+                trace!(
+                    target: "net::tx",
+                    peer_id=abbrev_hex(peer_id.as_ref()),
+                    hashes=abbrev_hex_hash_list(&hashes),
+                    "requesting buffered hashes from idle peer"
+                );
+
+                // request the buffered missing transactions
+                let metrics = &self.metrics;
+                if let Some(failed_to_request_hashes) =
+                    self.transaction_fetcher.request_transactions_from_peer(hashes, peer, || {
+                        metrics.egress_peer_channel_full.increment(1)
+                    })
+                {
+                    debug!(target: "net::tx",
+                        peer_id=abbrev_hex(peer_id.as_ref()),
+                        hashes=%abbrev_hex_hash_list(&failed_to_request_hashes),
+                        "failed sending request to peer's session, buffering hashes"
+                    );
+
+                    self.transaction_fetcher.buffer_hashes(failed_to_request_hashes, Some(peer_id));
+                    return
+                }
+            }
+        }
+    }
+
+    /// Returns any idle peer for the given hash. Writes peer IDs of any ended sessions to buffer
+    /// passed as parameter.
+    fn get_idle_peer_for(
+        &self,
+        hash: TxHash,
+        ended_sessions_buf: &mut Vec<PeerId>,
+    ) -> Option<PeerId> {
+        let (_, peers) = self.transaction_fetcher.unknown_hashes.peek(&hash)?;
+
+        for &peer_id in peers.iter() {
+            if self.transaction_fetcher.is_idle(peer_id) {
+                if self.peers.contains_key(&peer_id) {
+                    return Some(peer_id)
+                } else {
+                    ended_sessions_buf.push(peer_id);
+                }
+            }
+        }
+
+        None
+    }
+
+    /// Returns any idle peer for any buffered unknown hash, and writes that hash to the request
+    /// buffer passed as parameter.
+    fn pop_any_idle_peer(&mut self, hashes: &mut Vec<TxHash>) -> Option<PeerId> {
+        let mut ended_sessions = vec![];
+        let mut buffered_hashes_iter = self.transaction_fetcher.buffered_hashes.iter();
+
+        let idle_peer = loop {
+            let Some(&hash) = buffered_hashes_iter.next() else { break None };
+
+            let idle_peer = self.get_idle_peer_for(hash, &mut ended_sessions);
+            for peer_id in ended_sessions.drain(..) {
+                let (_, peers) = self.transaction_fetcher.unknown_hashes.peek_mut(&hash)?;
+                _ = peers.remove(&peer_id);
+            }
+            if idle_peer.is_some() {
+                hashes.push(hash);
+                break idle_peer
+            }
```

Comment:
ah okay, so filling the request is actually a two step process.

first we find the first idle peer that's suitable for a buffered hash.
then in a next step in `fill_request_for_peer` by going over all buffered hashes and check if the selected peer is a fallback

Reply:
that's right. we won't always go over all buffered hashes, if the request is filled to the limit halfway through, we break the loop early...but I see now that is only true for eth66! will fix it